test-data/unit/check-protocols.test

```diff
@@ -0,0 +1,1798 @@
+    def meth(self) -> int:
+        pass
+class C:
+    def meth(self) -> int:
+        pass
+
+x: C
+y: P
+x = y # E: Incompatible types in assignment (expression has type "P", variable has type "C")
+[out]
+
+[case testIndependentProtocolSubtyping]
+from typing import Protocol
+
+class P1(Protocol):
+    def meth(self) -> int:
+        pass
+class P2(Protocol):
+    def meth(self) -> int:
+        pass
+
+x1: P1
+x2: P2
+
+x1 = x2
+x2 = x1
+
+def f1(x: P1) -> None: pass
+def f2(x: P2) -> None: pass
+
+f1(x2)
+f2(x1)
+[out]
+
+[case testNoneDisablesProtocolImplementation]
+from typing import Protocol
+
+class MyHashable(Protocol):
+    def __my_hash__(self) -> int:
+        return 0
+
+class C:
+    __my_hash__ = None
+
+var: MyHashable = C()  # E: Incompatible types in assignment (expression has type "C", variable has type "MyHashable")
+[out]
+
+[case testNoneDisablesProtocolSubclassingWithStrictOptional]
+# flags: --strict-optional
+from typing import Protocol
+
+class MyHashable(Protocol):
+    def __my_hash__(self) -> int:
+        return 0
+
+class C(MyHashable):
+    __my_hash__ = None  # E: Incompatible types in assignment \
+(expression has type None, base class "MyHashable" defined the type as Callable[[MyHashable], int])
+[out]
+
+-- Semanal errors in protocol types
+-- --------------------------------
+
+[case testBasicSemanalErrorsInProtocols]
+from typing import Protocol, Generic, TypeVar, Iterable
+
+T = TypeVar('T', covariant=True)
+S = TypeVar('S', covariant=True)
+
+class P1(Protocol[T, T]): # E: Duplicate type variables in Generic[...] or Protocol[...]
+    def meth(self) -> T:
+        pass
+
+class P2(Protocol[T], Protocol[S]): # E: Only single Generic[...] or Protocol[...] can be in bases
+    def meth(self) -> T:
+        pass
+
+class P3(Protocol[T], Generic[S]): # E: Only single Generic[...] or Protocol[...] can be in bases
+    def meth(self) -> T:
+        pass
+
+class P4(Protocol[T]):
+    attr: Iterable[S] # E: Invalid type "__main__.S"
+
+class P5(Iterable[S], Protocol[T]): # E: If Generic[...] or Protocol[...] is present it should list all type variables
+    def meth(self) -> T:
+        pass
+[out]
+
+[case testProhibitSelfDefinitionInProtocols]
+from typing import Protocol
+
+class P(Protocol):
+    def __init__(self, a: int) -> None:
+        self.a = a # E: Protocol members cannot be defined via assignment to self \
+                   # E: "P" has no attribute "a"
+
+class B: pass
+class C:
+    def __init__(self, a: int) -> None:
+        pass
+
+x: P
+x = B()
+# The above has an incompatible __init__, but mypy ignores this for nominal subtypes?
+x = C(1)
+
+class P2(Protocol):
+    a: int
+    def __init__(self) -> None:
+        self.a = 1
+
+class B2(P2):
+    a: int
+
+x2: P2 = B2()  # OK
+[out]
+
+[case testProtocolsInTypingExtensions]
+from typing_extensions import Protocol, runtime
+
+@runtime
+class P(Protocol):
+    def meth(self) -> int:
+        pass
+
+x: object
+if isinstance(x, P):
```

Comment:
Yes, the error is ``Only @runtime protocols can be used with instance and class checks``